pkg/scheduler/framework/plugins/volumezone/volume_zone.go

```diff
@@ -116,16 +120,31 @@ func (pl *VolumeZone) PreFilter(ctx context.Context, cs *framework.CycleState, p
 	return nil, nil
 }
 
-func (pl *VolumeZone) getPVbyPod(ctx context.Context, pod *v1.Pod) ([]pvTopology, *framework.Status) {
-	logger := klog.FromContext(ctx)
+// checkStorageClassVolumeBindingModeIsWait confirms whether storageClass's volumeBindingMode is VolumeBindingWaitForFirstConsumer or not.
+func (pl *VolumeZone) checkStorageClassVolumeBindingModeIsWait(scName string) (bool, *framework.Status) {
+	if len(scName) == 0 {
+		return false, framework.NewStatus(framework.UnschedulableAndUnresolvable, "PersistentVolumeClaim had no pv name and storageClass name")
+	}
+	class, err := pl.scLister.Get(scName)
+	if s := getErrorAsStatus(err); !s.IsSuccess() {
+		return false, s
+	}
+	if class.VolumeBindingMode == nil {
+		return false, framework.NewStatus(framework.UnschedulableAndUnresolvable, fmt.Sprintf("VolumeBindingMode not set for StorageClass %q", scName))
+	}
+	if *class.VolumeBindingMode == storage.VolumeBindingWaitForFirstConsumer {
+		// Skip unbound volumes
+		return true, nil
+	}
+	return false, framework.NewStatus(framework.UnschedulableAndUnresolvable, "PersistentVolume had no name")
+}
+
+// getPVbyPod gets PVTopology from pod
+func (pl *VolumeZone) getPVbyPod(logger klog.Logger, pod *v1.Pod) ([]pvTopology, *framework.Status) {
 	podPVTopologies := make([]pvTopology, 0)
 
-	for i := range pod.Spec.Volumes {
-		volume := pod.Spec.Volumes[i]
-		if volume.PersistentVolumeClaim == nil {
-			continue
-		}
-		pvcName := volume.PersistentVolumeClaim.ClaimName
+	pvcNames := pl.getPersistentVolumeClaimNameFromPod(pod)
+	for _, pvcName := range pvcNames {
 		if pvcName == "" {
 			return nil, framework.NewStatus(framework.UnschedulableAndUnresolvable, "PersistentVolumeClaim had no name")
 		}
```

Comment:
GitHub doesn't allow me to comment L151 😓 

```go
pvc, err := pl.pvcLister.PersistentVolumeClaims(pod.Namespace).Get(pvcName)
if s := getErrorAsStatus(err); !s.IsSuccess() {
			return nil, s
}
```

What kind of error could this `Get` get? If, for example, it could get some temporal errors due to network failure or something, we should return error from isSchedulableAfterNodeChange when facing this error.

Reply:
@sanposhiho 
If this `Get` fails, a `NotFoundError` is returned.

In the `getErrorAsStatus` function, we distinguish whether the error was due to 'not found' or something else (like a network failure?). However, in both cases, it's not considered a success, so an error is returned here.

Does it make sense?

You can refer to past PRs and issues.
https://github.com/kubernetes/kubernetes/issues/102305
https://github.com/kubernetes/kubernetes/pull/102465/files

Reply:
> If this Get fails, a NotFoundError is returned.

You mean this Get only returns not found errors?

Reply:
My previous reply might have been incorrect.
Seems that Get only returns a "not found" error, so `getErrorAsStatus` only returns an `UnschedulableAndUnresolvable` Status or nil.


1 Get only returns a "not found" error.
https://github.com/kubernetes/client-go/blob/master/listers/core/v1/persistentvolumeclaim.go#L89-L99
```
// Get retrieves the PersistentVolumeClaim from the indexer for a given namespace and name.
func (s persistentVolumeClaimNamespaceLister) Get(name string) (*v1.PersistentVolumeClaim, error) {
	obj, exists, err := s.indexer.GetByKey(s.namespace + "/" + name)
	if err != nil {
		return nil, err
	}
	if !exists {
		return nil, errors.NewNotFound(v1.Resource("persistentvolumeclaim"), name)
	}
	return obj.(*v1.PersistentVolumeClaim), nil
}
```

https://github.com/kubernetes/client-go/blob/30eba26adb82c17a65cb10dbd9534c742a496404/tools/cache/store.go#L255-L258
```
// GetByKey returns the request item, or exists=false.
// GetByKey is completely threadsafe as long as you treat all items as immutable.
func (c *cache) GetByKey(key string) (item interface{}, exists bool, err error) {
	item, exists = c.cacheStorage.Get(key)
	return item, exists, nil
}
```

2 getErrorAsStatus returns an `UnschedulableAndUnresolvable` Status in the case of a "not found" error. For other errors, it returns an "Error" Status (however Get only returns "not found".)
https://github.com/kubernetes/kubernetes/blob/ae2e0c00b5d52571bf604aefa4f4efebb0afd4bc/pkg/scheduler/framework/plugins/volumezone/volume_zone.go#L265-L273
```
func getErrorAsStatus(err error) *framework.Status {
	if err != nil {
		if apierrors.IsNotFound(err) {
			return framework.NewStatus(framework.UnschedulableAndUnresolvable, err.Error())
		}
		return framework.AsStatus(err)
	}
	return nil
}
```

> If, for example, it could get some temporal errors due to network failure or something, we should return error from isSchedulableAfterNodeChange when facing this error.

Reply:
> If, for example, it could get some temporal errors due to network failure or something, we should return error from isSchedulableAfterNodeChange when facing this error.

@sanposhiho 
So do you mean that I should return error here instead of nil?
https://github.com/kubernetes/kubernetes/blob/8f296b932598c96aa383c4216cba57338f3c8c70/pkg/scheduler/framework/plugins/volumezone/volume_zone.go#L315-L318
```
	podPVTopologies, status := pl.getPVbyPod(logger, pod)
	if !status.IsSuccess() {
		return framework.QueueAfterBackoff, nil
	}
```

Reply:
Okay, yes. Let's return errors there, that'd be going to simpler.
Actually, errors from QHints are interpreted as `Queue` (= previous QueueAfterBackoff), so the result won't get changed anyway.

Reply:
Thanks, fixed